numbers/mixtureModel.go

```diff
@@ -0,0 +1,256 @@
+// starting values until convergence or maxResets. RunMixtureModel will store the results of the model in mm and will return whether the model
+// converged, and how many iterations it took to converge. If converged == false, the results in mm are meaningless.
+//
+// To reduce the number of allocations required for repeated use of RunMixtureModel, the input mixture model 'mm' can be reused between calls
+// with no modifications necessary.
+func RunMixtureModel(data []float64, k int, maxIterations int, maxResets int, mm *MixtureModel) (converged bool, iterationsRun int) {
+	if len(data) == 0 {
+		return
+	}
+	initMixtureModel(data, k, maxIterations, mm)
+	var resets int
+	var prevLogLikelihood float64
+
+	for iterationsRun = 0; resets < maxResets && !converged; iterationsRun++ {
+
+		// E step
+		prevLogLikelihood = mm.LogLikelihood
+		expectation(mm)
+		if iterationsRun > 2 && math.Abs(mm.LogLikelihood-prevLogLikelihood) < logProbEpsilon {
+			converged = true
+		}
+
+		// M step
+		maximization(mm)
+
+		// check to see if variance or weights are going to zero which is undesirable as it indicates that
+		// the model is stuck overfitting or trying to fit the data to fewer gaussians than intended.
+		for i := 0; i < mm.K; i++ {
+			switch {
+			case mm.Stdev[i] < 1e-04: // most likely caused by overfit of gaussian to a single data point
+				fallthrough
+			case mm.Weights[i] < 1e-02: // they asked for 2 gaussians, reset and see if you can get better answer
+				resets++
+				initMixtureModel(data, k, maxIterations, mm)
+				iterationsRun = 0
+				prevLogLikelihood = 0
+				converged = false
+				break
+			}
+		}
+
+		// if max iterations reached, reset with new values and try again
+		if iterationsRun == mm.MaxIter {
+			resets++
+			initMixtureModel(data, k, maxIterations, mm)
+			iterationsRun = 0
+			prevLogLikelihood = 0
+			converged = false
+		}
+	}
+	return
+}
+
+// initMixtureModel allocates and fills the MixtureModel struct. Avoids allocation if enough space is already present in input struct.
+func initMixtureModel(data []float64, k int, maxIterations int, mm *MixtureModel) {
+	if mm == nil {
+		mm = new(MixtureModel)
+	}
+	mm.MaxIter = maxIterations
+	mm.Data = data
+	mm.K = k
+	mm.LogLikelihood = math.MaxFloat64
+	if cap(mm.Means) >= k {
+		mm.Means = mm.Means[0:k]
+	} else {
+		mm.Means = make([]float64, k)
+	}
+	if cap(mm.Stdev) >= k {
+		mm.Stdev = mm.Stdev[0:k]
+	} else {
+		mm.Stdev = make([]float64, k)
+	}
+	if cap(mm.lamSigRatio) >= k {
+		mm.lamSigRatio = mm.lamSigRatio[0:k]
+	} else {
+		mm.lamSigRatio = make([]float64, k)
+	}
+	if cap(mm.logLamSigRatio) >= k {
+		mm.logLamSigRatio = mm.logLamSigRatio[0:k]
+	} else {
+		mm.logLamSigRatio = make([]float64, k)
+	}
+	if cap(mm.work) >= k {
+		mm.work = mm.work[0:k]
+	} else {
+		mm.work = make([]float64, k)
+	}
+
+	// TODO smarter initial guess for mean and variance (k-means/PCA)
+	for i := range mm.Means {
```

Comment:
Most k-means implementations (which is very similar to Gaussian mixture models) pick the location of random data points as the starting means.

Reply:
Good idea. Added